tests/model_fields/test_foreignkey.py

```diff
@@ -147,3 +148,21 @@ class Related(models.Model):
         )
         with self.assertRaisesMessage(FieldError, msg):
             Related._meta.get_field('child').related_fields
+
+    @isolate_apps('model_fields')
+    def test_distinct_choice_limit(self):
+        a = Foo.objects.create(a='a', d=Decimal("-1"))
+        b = Foo.objects.create(a='b', d=Decimal("1"))
+
+        Bar.objects.create(b='ah', a=a)
+        Bar.objects.create(b='aha', a=a)
+        Bar.objects.create(b='bla', a=b)
+
+        class BazForm(forms.ModelForm):
+            class Meta:
+                model = Baz
+                fields = ["a", "foo"]
+
+        form = BazForm()
+        baz_objects = [baz_object.data['label'] for baz_object in form['foo']]
+        self.assertEqual(len(baz_objects), len(set(baz_objects)))
```

Comment:
These aren't `Baz` instances, but `Foo` ones. Also, generating the expected values using almost the same code as the code under test seems circular. I'd prefer to test explicitly for the `Foo`'s we expect in the form. Something like:
```python
# Check we get `a` exactly once as an option even though it
# has multiple `Bar` objects where `Bar.b.startswith('a')`
non_empty_foo_choices = [option.data['value'].instance for option in form['foo'][1:]]
self.assertEqual(non_empty_foo_choices, [a])
```

Reply:
Thank you. This is really helpful. I had not realize the naming was incorrect. Plus the solution is more direct to what we are testing.